sonic_data_client/non_db_client.go

```diff
@@ -332,7 +333,107 @@ func (c *NonDbClient) String() string {
 
 // To be implemented
 func (c *NonDbClient) StreamRun(q *queue.PriorityQueue, stop chan struct{}, w *sync.WaitGroup, subscribe *gnmipb.SubscriptionList) {
-	return
+	c.w = w
+	defer c.w.Done()
+	c.q = q
+
+	validatedSubs := make(map[*gnmipb.Subscription]time.Duration)
+
+	// Validate all subs
+	for _, sub := range subscribe.GetSubscription() {
+		gnmiPath := sub.GetPath()
+		_, ok := c.path2Getter[gnmiPath]
+		if !ok {
+			log.V(3).Infof("Cannot find getter for the path: %v", gnmiPath)
+			continue
+		}
+
+		subMode := sub.GetMode()
+		if subMode != gnmipb.SubscriptionMode_SAMPLE {
+			putFatalMsg(c.q, fmt.Sprintf("Unsupported subscription mode: %v.", subMode))
+			return
+		}
+
+		interval, err := validateSampleInterval(sub)
+		if err != nil {
+			putFatalMsg(c.q, err.Error())
+			return
+		}
+
+		validatedSubs[sub] = interval
+	}
+
+	if len(validatedSubs) == 0 {
+		log.V(3).Infof("No valid sub for stream subsription.")
+		return
+	}
+
+	for sub := range validatedSubs {
+		gnmiPath := sub.GetPath()
+		getter, _ := c.path2Getter[gnmiPath]
+		runGetterAndSend(c, gnmiPath, getter)
+	}
+
+	c.q.Put(Value{
+		&spb.Value{
+			Timestamp:    time.Now().UnixNano(),
+			SyncResponse: true,
+		},
+	})
+
+	// Start a GO routine for each sub as they might have different intervals
+	for sub, interval := range validatedSubs {
+		go sampledStream(c, stop, sub, interval)
+	}
+
+	log.V(1).Infof("Started non-db sampling routines for %s ", c)
```

Comment:
Is this needed here? Line 396 got similar log statement?

Reply:
There is a slight difference. line 389 is for the whole subscription request. However, line 396 is for the individual subs (e.g. for each query path).
E.g. the subcriptionlist below request would produce the following logs:
```
SAMPLE proc/loadavg
SAMPLE proc/vmstat
```

Logs:
```
I1120 12:55:33.577640 3664053 non_db_client.go:389] Started non-db sampling routines for NonDbClient Prefix OTHERS  sendMsg 0, recvMsg 0 
I1120 12:55:33.577666 3664053 non_db_client.go:396] Starting sampling routine sub: 'path:{elem:{name:"proc"} elem:{name:"loadavg"}} mode:SAMPLE sample_interval:30000000000 heartbeat_interval:60000000000' client: 'NonDbClient Prefix OTHERS  sendMsg 0, recvMsg 0'
I1120 12:55:33.577728 3664053 non_db_client.go:396] Starting sampling routine sub: 'path:{elem:{name:"proc"} elem:{name:"vmstat"}} mode:SAMPLE sample_interval:30000000000 heartbeat_interval:60000000000' client: 'NonDbClient Prefix OTHERS  sendMsg 0, recvMsg 0'
```